### PR TITLE
Collection of Resource Metadata implementation 

### PR DESCRIPTION
**Pull request by soyuka**
This is an implementation of the Resource Metadata as defined in https://github.com/api-platform/core/blob/main/docs/adr/0002-resource-definition.md.

Things that changes (maintaining an upgrade path with deprecations):

- new namespace `ApiPlatform\Metadata` instead of `ApiPlatform\Core\Metadata`, for example `ApiPlatform\Metadata\ApiResource`
- deprecation of `ApiPlatform\Core\Annotation`
- `ResourceMetadataFactoryInterface` is deprecated in favor of `ResourceMetadataCollectionFactoryInterface` 
- each `ApiPlatform\Metadata\ApiResource` holds various `ApiPlatform\Metadata\Operation[]`
-`item` and `collection` prefixes for operations are deprecated
- ApiPlatform\Metadata\GraphQl follows these conventions (a Subscription operation is available and isn't hidden behind an `update` Mutation anymore) 
- GraphQl interfaces got simplified (beeing `@experimental`)
- tests have a new location at `tests/Core` and some are flagged `@legacy`, the `@legacy` tests will need an update and to be moved inside `tests/` (we will remove `tests/Core` when working on 3.0)
- new interface for `IriConverter` that adds an `operationName` 
- new interface for `IdentifiersExtractor` (should probably be tagged `@internal`)
- new `ApiPlatform\State\ProviderInterface` and `ApiPlatform\State\ProcessorInterface` with a bc-layer

Note that on this branch the legacy `ResourceMetadata` still works and is being transformed to a `ResourceMetadataCollection`.

A new configuration is available to keep old services (`IriConverter`, `IdentifiersExtractor` and `OpenApiFactory`):

```
    metadata_backward_compatibility_layer: false
```

Because things moved, focus the review on the following commits:

```
feat(metadata): Resource Collection
feat(graphql): graphql resource metadata collection 
``` 

TODO:

- MongoDB
- Elasticsearch

### CODE REVIEW

File: src/Core/EventListener/WriteListener.php [hunk 1]
```diff
@@ -31,18 +36,32 @@
...
     public const OPERATION_ATTRIBUTE_KEY = 'write';
 
     private $dataPersister;
     private $iriConverter;
 
-    public function __construct(DataPersisterInterface $dataPersister, IriConverterInterface $iriConverter = null, ResourceMetadataFactoryInterface $resourceMetadataFactory = null, ResourceClassResolverInterface $resourceClassResolver = null)
+    public function __construct($dataPersister, $iriConverter = null, $resourceMetadataFactory = null, ResourceClassResolverInterface $resourceClassResolver = null)
     {
+        if ($dataPersister instanceof DataPersisterInterface) {
+            trigger_deprecation('api-platform/core', '2.7', sprintf('Use an implementation of "%s" instead of "%s".', DataPersisterInterface::class, ProcessorInterface::class));
```
Comment (alanpoulain) [inline src/Core/EventListener/WriteListener.php:52]: The classes should be inverted.

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ConfigurationTest.php [hunk 2]
```diff
@@ -224,9 +224,6 @@ private function runDefaultConfigTests(array $doctrineIntegrationsToLoad = ['orm
                 'backward_compatibility_layer' => true,
                 'swagger_ui_extra_configuration' => [],
             ],
-            'maker' => [
-                'enabled' => true,
-            ],
```
Comment (alanpoulain) [inline tests/Bridge/Symfony/Bundle/DependencyInjection/ConfigurationTest.php:229]: Should not be removed.

File: tests/Fixtures/TestBundle/Entity/JsonldContextDummy.php [hunk 3]
```diff
@@ -42,7 +42,7 @@ class JsonldContextDummy
      *     attributes={
      *         "jsonld_context"={
      *             "@id"="http://example.com/id",
-     *             "@type"="@id",
+     *             "@var"="@id",
```
Comment (soyuka) [inline tests/Fixtures/TestBundle/Entity/JsonldContextDummy.php:45]: revert

File: tests/Fixtures/app/AppKernel.php [hunk 4]
```diff
@@ -75,7 +74,6 @@ public function registerBundles(): array
             new WebProfilerBundle(),
             new FriendsOfBehatSymfonyExtensionBundle(),
             new FrameworkBundle(),
-            new MakerBundle(),
```
Comment (soyuka) [inline tests/Fixtures/app/AppKernel.php:78]: revert

File: tests/Hydra/JsonSchema/SchemaFactoryTest.php [hunk 5]
```diff
@@ -136,18 +136,4 @@ public function testSchemaTypeBuildSchema(): void
...
-
-        $this->assertNull($resultSchema->getRootDefinitionKey());
-        $this->assertArrayHasKey('properties', $resultSchema);
-        $this->assertArrayHasKey('hydra:view', $resultSchema['properties']);
-        $this->assertArrayHasKey('properties', $resultSchema['properties']['hydra:view']);
-        $this->assertArrayHasKey('hydra:first', $resultSchema['properties']['hydra:view']['properties']);
-        $this->assertArrayHasKey('hydra:last', $resultSchema['properties']['hydra:view']['properties']);
-        $this->assertArrayHasKey('hydra:previous', $resultSchema['properties']['hydra:view']['properties']);
-        $this->assertArrayHasKey('hydra:next', $resultSchema['properties']['hydra:view']['properties']);
-    }
```
Comment (soyuka) [inline tests/Hydra/JsonSchema/SchemaFactoryTest.php:152]: revert

File: tests/Bridge/Doctrine/EventListener/PublishMercureUpdatesListenerTest.php [hunk 6]
```diff
@@ -180,15 +180,12 @@ public function testPublishUpdateWithLegacySignature(): void
         $iriConverterProphecy->getIriFromItem($toDeleteExpressionLanguage, UrlGeneratorInterface::ABS_URL)->willReturn('http://example.com/dummy_friends/4')->shouldBeCalled();
         $iriConverterProphecy->getIriFromItem($toDeleteMercureOptions)->willReturn('/dummy_offers/5')->shouldBeCalled();
         $iriConverterProphecy->getIriFromItem($toDeleteMercureOptions, UrlGeneratorInterface::ABS_URL)->willReturn('http://example.com/dummy_offers/5')->shouldBeCalled();
-        $iriConverterProphecy->getIriFromItem($toDeleteMercureOptions, UrlGeneratorInterface::ABS_PATH)->willReturn('/dummy_offers/5')->shouldBeCalled();
-        $iriConverterProphecy->getIriFromItem($toDeleteMercureOptions, UrlGeneratorInterface::REL_PATH)->willReturn('./dummy_offers/5')->shouldBeCalled();
-        $iriConverterProphecy->getIriFromItem($toDeleteMercureOptions, UrlGeneratorInterface::NET_PATH)->willReturn('//example.com/dummy_offers/5')->shouldBeCalled();
```
Comment (alanpoulain) [inline tests/Bridge/Doctrine/EventListener/PublishMercureUpdatesListenerTest.php:185]: Should not be removed.

File: tests/Bridge/Doctrine/EventListener/PublishMercureUpdatesListenerTest.php [hunk 7]
```diff
@@ -180,15 +180,12 @@ public function testPublishUpdateWithLegacySignature(): void
...
         $resourceMetadataFactoryProphecy->create(DummyFriend::class)->willReturn(new ResourceMetadata(null, null, null, null, null, ['mercure' => ['private' => true, 'retry' => 10]]));
-        $resourceMetadataFactoryProphecy->create(DummyOffer::class)->willReturn(new ResourceMetadata(null, null, null, null, null, ['mercure' => ['topics' => ['http://example.com/custom_topics/1', '@=iri(object)', '@=iri(object, '.UrlGeneratorInterface::ABS_URL.')', '@=iri(object, '.UrlGeneratorInterface::ABS_PATH.')', '@=iri(object, '.UrlGeneratorInterface::REL_PATH.')', '@=iri(object, '.UrlGeneratorInterface::NET_PATH.')', '@=iri(object) ~ "/?topic=" ~ escape(iri(object))', '@=iri(object) ~ "/?topic=" ~ escape(iri(object, '.UrlGeneratorInterface::ABS_PATH.'))'], 'data' => 'mercure_custom_data', 'normalization_context' => ['groups' => ['baz']]]]));
+        $resourceMetadataFactoryProphecy->create(DummyOffer::class)->willReturn(new ResourceMetadata(null, null, null, null, null, ['mercure' => ['topics' => 'http://example.com/custom_topics/1', 'data' => 'mercure_custom_data', 'normalization_context' => ['groups' => ['baz']]]]));
```
Comment (alanpoulain) [inline tests/Bridge/Doctrine/EventListener/PublishMercureUpdatesListenerTest.php:188]: Should not be removed.

File: tests/Bridge/Doctrine/EventListener/PublishMercureUpdatesListenerTest.php [hunk 8]
```diff
@@ -235,7 +232,7 @@ public function testPublishUpdateWithLegacySignature(): void
...
-        $this->assertSame(['http://example.com/dummies/1', 'http://example.com/dummies/2', 'http://example.com/custom_topics/1', 'http://example.com/dummy_offers/5', 'http://example.com/dummy_offers/5', '/dummy_offers/5', './dummy_offers/5', '//example.com/dummy_offers/5', 'http://example.com/dummy_offers/5/?topic=http%3A%2F%2Fexample.com%2Fdummy_offers%2F5', 'http://example.com/dummy_offers/5/?topic=%2Fdummy_offers%2F5', '/dummies/1', '/users/3', 'http://example.com/dummies/3', 'http://example.com/dummy_friends/4', 'http://example.com/custom_topics/1', 'http://example.com/dummy_offers/5', 'http://example.com/dummy_offers/5', '/dummy_offers/5', './dummy_offers/5', '//example.com/dummy_offers/5', 'http://example.com/dummy_offers/5/?topic=http%3A%2F%2Fexample.com%2Fdummy_offers%2F5', 'http://example.com/dummy_offers/5/?topic=%2Fdummy_offers%2F5'], $topics);
+        $this->assertSame(['http://example.com/dummies/1', 'http://example.com/dummies/2', 'http://example.com/custom_topics/1', '/dummies/1', '/users/3', 'http://example.com/dummies/3', 'http://example.com/dummy_friends/4', 'http://example.com/custom_topics/1'], $topics);
```
Comment (alanpoulain) [inline tests/Bridge/Doctrine/EventListener/PublishMercureUpdatesListenerTest.php:235]: Should not be changed.

File: tests/Bridge/Doctrine/Orm/Filter/SearchFilterTest.php [hunk 9]
```diff
@@ -343,7 +343,7 @@ public function testNotPassingIdentifiersExtractor()
         $requestStack->push(Request::create('/api/dummies', 'GET', []));
         $iriConverterProphecy = $this->prophesize(IriConverterInterface::class);
         $iriConverter = $iriConverterProphecy->reveal();
-        $propertyAccessor = static::$kernel->getContainer()->get('test.property_accessor');
+        $propertyAccessor = self::$kernel->getContainer()->get('test.property_accessor');
```
Comment (alanpoulain) [inline tests/Bridge/Doctrine/Orm/Filter/SearchFilterTest.php:346]: `static` should be used.

File: tests/Bridge/Doctrine/Orm/Filter/SearchFilterTest.php [hunk 10]
```diff
@@ -593,7 +593,7 @@ protected function buildSearchFilter(ManagerRegistry $managerRegistry, ?array $p
         });
 
         $iriConverter = $iriConverterProphecy->reveal();
-        $propertyAccessor = static::$kernel->getContainer()->get('test.property_accessor');
+        $propertyAccessor = self::$kernel->getContainer()->get('test.property_accessor');
```
Comment (alanpoulain) [inline tests/Bridge/Doctrine/Orm/Filter/SearchFilterTest.php:596]: `static` should be used.

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php [hunk 11]
```diff
@@ -1356,8 +1380,6 @@ private function getBaseContainerBuilderProphecyWithoutDefaultMetadataLoading(ar
             'api_platform.json_schema.schema_factory',
             'api_platform.listener.view.validate',
             'api_platform.listener.view.validate_query_parameters',
-            'api_platform.maker.command.data_persister',
-            'api_platform.maker.command.data_provider',
```
Comment (alanpoulain) [inline tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php:1360]: Should not be removed.

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php [hunk 12]
```diff
@@ -1388,14 +1410,14 @@ private function getBaseContainerBuilderProphecyWithoutDefaultMetadataLoading(ar
             'api_platform.metadata.resource.name_collection_factory.annotation',
             'api_platform.metadata.resource.name_collection_factory.yaml',
             'api_platform.metadata.subresource.metadata_factory.annotation',
-            'api_platform.argument_resolver.payload',
```
Comment (alanpoulain) [inline tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php:1391]: Should not be removed.

File: src/Metadata/Operation.php [hunk 13]
```diff
@@ -0,0 +1,1113 @@
...
+     * @param string          $securityMessage                https://api-platform.com/docs/core/security/#configuring-the-access-control-error-message
+     * @param string          $securityPostDenormalize        https://api-platform.com/docs/core/security/#executing-access-control-rules-after-denormalization
+     * @param string          $securityPostDenormalizeMessage https://api-platform.com/docs/core/security/#configuring-the-access-control-error-message
+     * @param bool            $compositeIdentifier
+     * @param GraphQl|null    $graphQl
+     * @param bool            $read                           https://api-platform.com/docs/core/events/#the-event-system
+     * @param bool            $deserialize                    https://api-platform.com/docs/core/events/#the-event-system
+     * @param bool            $validate                       https://api-platform.com/docs/core/events/#the-event-system
+     * @param bool            $write                          https://api-platform.com/docs/core/events/#the-event-system
+     * @param bool            $serialize                      https://api-platform.com/docs/core/events/#the-event-system
```
Comment (alanpoulain) [inline src/Metadata/Operation.php:92]: Missing `$exceptionToStatus`?

File: src/Metadata/Resource/DeprecationMetadataTrait.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+        } elseif ('validation_groups' === $key) {
+            @trigger_error('The "validation_groups" is deprecated in 2.7 and will be renamed to "validation_context" having an array with a "groups" key in 3.0.', \E_USER_DEPRECATED);
+            $key = 'validation_context';
+            $value = ['groups' => \is_array($value) ? $value : [$value]];
+        } elseif ('access_control' === $key) {
+            $key = 'security';
+            @trigger_error('The "access_control" option is deprecated in 2.7 and will be renamed to "security" in 3.0.', \E_USER_DEPRECATED);
+        } elseif ('path' === $key) {
+            $key = 'uri_template';
+            @trigger_error('The "path" option is deprecated in 2.7 and will be renamed to "uri_tempalte" in 3.0.', \E_USER_DEPRECATED);
```
Comment (alanpoulain) [inline src/Metadata/Resource/DeprecationMetadataTrait.php:47]:
```suggestion
            @trigger_error('The "path" option is deprecated in 2.7 and will be renamed to "uri_template" in 3.0.', \E_USER_DEPRECATED);
```

File: src/Metadata/Operation.php [hunk 15]
```diff
@@ -0,0 +1,1113 @@
...
+     * @param array|string    $inputFormats                   https://api-platform.com/docs/core/content-negotiation/#configuring-formats-for-a-specific-resource-or-operation
+     * @param array|string    $outputFormats                  https://api-platform.com/docs/core/content-negotiation/#configuring-formats-for-a-specific-resource-or-operation
+     * @param array           $identifiers
+     * @param array           $links
+     * @param string          $routePrefix                    https://api-platform.com/docs/core/operations/#prefixing-all-routes-of-all-operations
+     * @param string          $routeName
+     * @param array           $defaults
+     * @param array           $requirements
+     * @param array           $options
+     * @param int             $referenceType
```
Comment (alanpoulain) [inline src/Metadata/Operation.php:44]: Should probably be removed (prefer `urlGenerationStrategy` instead).

**Comment by Chris53897** [comment]
Hi. Can you please give us a little bit infos about this pr? 711 Files changed looks like a great step forward to Apip 3.0.

**Comment by soyuka** [comment]
@Chris53897 it's an implementation of https://github.com/api-platform/core/blob/main/docs/adr/0002-resource-definition.md !

File: src/Core/OpenApi/Factory/OpenApiFactory.php [hunk 16]
```diff
@@ -0,0 +1,533 @@
...
+        $baseUrl = $context[self::BASE_URL] ?? '/';
+        $contact = null === $this->openApiOptions->getContactUrl() || null === $this->openApiOptions->getContactEmail() ? null : new Model\Contact($this->openApiOptions->getContactName(), $this->openApiOptions->getContactUrl(), $this->openApiOptions->getContactEmail());
+        $license = null === $this->openApiOptions->getLicenseName() ? null : new Model\License($this->openApiOptions->getLicenseName(), $this->openApiOptions->getLicenseUrl());
+        $info = new Model\Info($this->openApiOptions->getTitle(), $this->openApiOptions->getVersion(), trim($this->openApiOptions->getDescription()), $this->openApiOptions->getTermsOfService(), $contact, $license);
+        $servers = '/' === $baseUrl || '' === $baseUrl ? [new Model\Server('/')] : [new Model\Server($baseUrl)];
+        $paths = new Model\Paths();
+        $links = [];
+        $schemas = new \ArrayObject();
+
+        foreach ($this->resourceNameCollectionFactory instanceof LegacyResourceNameCollectionFactoryInterface ? $this->resourceNameCollectionFactory->create(true) : $this->resourceNameCollectionFactory->create() as $resourceClass) {
```
Comment (vincentchalamon) [inline src/Core/OpenApi/Factory/OpenApiFactory.php:93]: Interface `LegacyResourceNameCollectionFactoryInterface` does not exist anywhere in the project. I guess you meant the legacy `ResourceNameCollectionFactoryInterface`, but I can't find it anywhere...

What did you mean here?

File: src/Core/Bridge/Symfony/Messenger/DataTransformer.php [hunk 17]
```diff
@@ -52,6 +60,17 @@ public function supportsTransformation($data, string $to, array $context = []):
             return false;
         }
 
+        if ($this->resourceMetadataFactory instanceof ResourceMetadataCollectionFactoryInterface) {
+            try {
+                $resourceMetadataCollection = $this->resourceMetadataFactory->create($context['resource_class'] ?? $this->getObjectClass($data));
```
Comment (vincentchalamon) [inline src/Core/Bridge/Symfony/Messenger/DataTransformer.php:69]: `getObjectClass` requires an object, which is impossible according to line 56.

File: src/Core/Bridge/Symfony/Bundle/ArgumentResolver/PayloadArgumentResolver.php [hunk 18]
```diff
@@ -68,10 +65,24 @@ public function resolve(Request $request, ArgumentMetadata $argument): \Generato
     private function getExpectedInputClass(Request $request): ?string
     {
         $attributes = RequestAttributesExtractor::extractAttributes($request);
-
-        if (!$this->isRequestToDeserialize($request, $attributes)) {
-            return null;
-        }
+        dd($attributes);
```
Comment (vincentchalamon) [inline src/Core/Bridge/Symfony/Bundle/ArgumentResolver/PayloadArgumentResolver.php:69]: Don't forget to remove this line, and uncomment or solve the following block

**Review by vincentchalamon — CHANGES_REQUESTED** [review]

**Comment by soyuka** [comment]
still needs some behat fixing, I've updated the description

**Pull request merged**